katsdptelstate/aio/redis.py

```diff
@@ -87,23 +87,42 @@ def __init__(self, client: aioredis.Redis) -> None:
                 'katsdptelstate', 'lua_scripts/{}.lua'.format(script_name))
             self._scripts[script_name] = _Script(script)
 
+    async def _execute(self, call: Callable[..., Awaitable], *args, **kwargs) -> Any:
+        """Execute a redis command, retrying if the connection died.
+
+        This handles cases like an idle connection being silently closed,
+        the server restarting etc. If there is a connection error, a fresh
+        connection is made and one more attempt is made.
+
+        Since there is no way to tell if the original attempt actually
+        succeeded, the command much be idempotent.
+        """
+        try:
+            return await call(*args, **kwargs)
+        except (ConnectionError, aioredis.ConnectionClosedError) as exc:
+            # Closes all connections in the pool to ensure that we get a
+            # fresh connection.
+            logger.warning('redis connection error (%s), trying again', exc)
+            await self.client.connection.clear()
+            return await call(*args, **kwargs)
+
     async def _call(self, script_name: str, *args, **kwargs) -> Any:
-        return await self._scripts[script_name](self.client, *args, **kwargs)
+        return await self._execute(self._scripts[script_name], self.client, *args, **kwargs)
```

Comment:
Given the idempotent requirement is it safe to use _execute in the generic _call function ?

Reply:
All the scripts are idempotent (other than that there would be a duplicate pubsub message, but that's pretty harmless). Should I document an idempotency requirement in `_call`?

Reply:
I think so since all our *current* scripts are, but who knows what the future holds...

Reply:
Done.